Use the audit ID of a request for better correlation

tkashem opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Use the audit ID of a request for better correlation:
- If an audit ID is not provided by the caller, generate one when we receive the request. This audit ID will be used to audit the request. 
- When we forward the request to the aggregated server, set the audit ID in the new request header. This allows audit logs from aggregated apiservers to be correlated with the kube-apiserver
- Use the audit ID in the current `tracer` we have
- Use the audit ID in `httplog`
- When a request panics, log an error with the audit ID.

#### Which issue(s) this PR fixes:

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

tkashem commented:
/retest

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:
```diff
@@ -0,0 +1,60 @@
+			auditID = newAuditIDFunc()
+		}
+
+		r = r.WithContext(request.WithAuditID(ctx, types.UID(auditID)))
```
The audit ID is stored in the request context here. When a caller specifies an audit ID, it will also be in the request header `Audit-ID`. 
Another alternative is to add the generated audit id to the request header. But we treat the `http.Request` object as immutable.

sttts replied:
This goes into direction of tracing. It rings by secrurity bell: if you want to inject this, you can fake it to hide yourself, and/or to put 8kb of data in a number of places (logs, audit events).

tkashem commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -762,8 +763,10 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
+	handler = httplog.WithLogging(handler, httplog.DefaultStacktracePred)
 	handler = genericapifilters.WithRequestReceivedTimestamp(handler)
 	handler = genericfilters.WithPanicRecovery(handler, c.RequestInfoResolver)
+	handler = genericapifilters.WithAuditID(handler)
```
It's before `WithPanicRecovery` - this is so we can access the audit ID from the context when we log error message inside the `WithPanicRecovery`  handler. 
If `WithAuditID` panics, `net/http2` will catch it and then log the panic with stack trace.

sttts replied:
sounds fine

tkashem commented:
/retest

tkashem commented:
/retest

fedebongio commented:
/triage accepted

tkashem commented:
/retest

tkashem commented:
/retest

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:
```diff
@@ -0,0 +1,60 @@
+/*
```
nit: auditid.go

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_audit_id.go:
```diff
@@ -0,0 +1,60 @@
+		r = r.WithContext(request.WithAuditID(ctx, types.UID(auditID)))
+
+		// Should we echo the Audit-ID for all requests?
+		// w.Header().Set(auditinternal.HeaderAuditID, auditID)
```
I remember some discussion about that long long ago. @soltysh @deads2k maybe have better memory

sttts replied:
Have you checked the audit enhancement? (if that exists)

tkashem replied:
Going through the comments, it looks like the intention was to set it for all requests. So I think moving it here makes sense.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/audit_id.go:
```diff
@@ -0,0 +1,64 @@
+	return info, ok
+}
+
+func GetVerbURIAndAuditID(req *http.Request) []interface{} {
```
what is this?

tkashem replied:
I removed it and doing it inline now.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -51,16 +50,15 @@ func WithPanicRecovery(handler http.Handler, resolver request.RequestInfoResolve
 			// in case the rate limit delays it.  If you outrun the rate for this one timed out requests, something has gone
 			// seriously wrong with your server, but generally having a logging signal for timeouts is useful.
-			runtime.HandleError(fmt.Errorf("timeout or abort while handling: %v %q", req.Method, req.URL.Path))
+			runtime.HandleError(fmt.Errorf("timeout or abort while handling: %s", request.GetVerbURIAndAuditIDString(req)))
```
What do we buy with this untyped way of writing this?

tkashem replied:
having a consistent key names (with structured logging). But I remove it and doing it inline now.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:
```diff
@@ -102,16 +109,20 @@ func respLoggerFromContext(req *http.Request) *respLogger {
 
-// newLogged turns a normal response writer into a logged response writer.
-func newLogged(req *http.Request, w http.ResponseWriter) *respLogger {
+func newLoggedWithStartTime(req *http.Request, w http.ResponseWriter, startTime time.Time) *respLogger {
```
unrelated refactor?

tkashem replied:
Yes, will open a follow up PR. the httplogger should use the `RequestReceivedTimestamp` from the request context. Then we will have both audit and httplog print the same exact timestamp.

sttts commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -152,6 +154,9 @@ func NewRequest(c *RESTClient) *Request {
 	}
+
+	// generate an audit ID
+	r.SetHeader("Audit-ID", uuid.New().String())
```
I feel like this whole cross-component topic makes the change much bigger, i.e. crossing the line to need an enhancement.

tkashem replied:
yes, I removed the client-go changes. I will create a new KEP with the proposed client-go changes.

tkashem committed:
apiserver: manage audit ID associated with a request

Manage the audit ID early in the request handling logic so that it can
be used by different layers to improve correlation.
- If the caller does not specify a value for Audit-ID in the request
  header, we generate a new audit ID
- If a user specified Audit-ID is too large, we truncate it
- We echo the Audit-ID value to the caller via the response
  Header 'Audit-ID'

tkashem committed:
apiserver: improve correlation by using the audit ID

- when we forward the request to the aggregated server, set the audit
  ID in the new request header. This allows audit logs from aggregated
  apiservers to be correlated with the kube-apiserver.
- use the audit ID in the current tracer
- use the audit ID in httplog
- when a request panics, log an error with the audit ID.

tkashem commented:
/retest

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -772,16 +773,16 @@ func TestAuditIDHttpHeader(t *testing.T) {
 			"",
 			auditinternal.LevelNone,
-			false,
+			true,
```
so this explicitly changed, right? Is the argument that we want to have an id in any case, even when we don't write audit events?

sttts replied:
A comment about the high level reason would be helpful

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:
```diff
@@ -0,0 +1,80 @@
+		if len(auditID) > 36 {
+			attachAuditIDToCtx = true
+			auditID = auditID[0:36]
+		}
```
this is new. How do we know we don't break anybody? Why 36 chars?

tkashem replied:
- yes, since we are copying the audit ID into the request context I wanted to a put a defensive guard to avoid copying large amount of data into the context. 
- Also, printing a very large audit ID is going to make reading the log very cumbersome. The client can still search with the first 36 characters.
- I chose 36, because the formatted UID we generate today is 36 characters long. I am fine increasing it but would still like us to have a limit.
- Well client-go does not generate any audit ID today. I don't think it breaks anyone today. The caller can still search with the first N characters.

p0lyn0mial replied:
Does it mean the input hasn't been sanitized before? so clients could send potentially very large inputs.
AFAIK the HTTP spec doesn't specify a limit for Headers.

tkashem replied:
for now, I am saving the audit ID as is in the context. But when we are logging (tracing and httplog) I am only showing the first N(64) character.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go:
```diff
@@ -0,0 +1,65 @@
+		return auditID, ok
+	}
+
+	if auditID := r.Header.Get(auditinternal.HeaderAuditID); len(auditID) > 0 {
```
I find all the logic starting from here irritating in this context. Would prefer exactly one place where we have header logic.

tkashem replied:
@sttts This is an optimization - If the audit ID is in the request header and we can use it as is then avoid copying it to the context. do you want me to go back to `always copy the audit ID in the request context`?

tkashem replied:
@sttts I removed the optimization, the context always has an audit ID.

tkashem commented:
/retest

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -762,8 +763,10 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
+	handler = httplog.WithLogging(handler, httplog.DefaultStacktracePred)
```
why have we moved it here?

tkashem replied:
good catch, this was an artifact for another follow up PR. I remove it.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:
```diff
@@ -157,12 +158,14 @@ func (rl *respLogger) Addf(format string, data ...interface{}) {
 
 func (rl *respLogger) LogArgs() []interface{} {
 	latency := time.Since(rl.startTime)
+	auditID := request.GetAuditID(rl.req)
```
it cannot be empty, right ?

tkashem replied:
if `WithAuditID` is wired, it cannot be empty.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:
```diff
@@ -51,16 +50,15 @@ func WithPanicRecovery(handler http.Handler, resolver request.RequestInfoResolve
+			runtime.HandleError(fmt.Errorf("timeout or abort while handling: method=%v URI=%q audit-ID=%q", req.Method, req.RequestURI, request.GetAuditID(req)))
 			return
 		}
 		http.Error(w, "This request caused apiserver to panic. Look in the logs for details.", http.StatusInternalServerError)
```
should we change the msg to contain the correlation ID as well? Users won't see it by default only with increased log level.

tkashem replied:
well, the audit ID is set in the response header. Also, for when net/http catches the panic it will close the connection or reset the stream for http2. So the user doesn't see the http response, I think.

tkashem commented:
/retest

tkashem commented:
Hi @liggitt can you please take a look when you have some time? 

/assign @liggitt

tkashem commented:
/assign @sttts

soltysh commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/helpers.go:
```diff
@@ -73,3 +74,17 @@ func (lazy *lazyAccept) String() string {
 }
+
+// lazyAuditID implements String() string and it will retrieve the audit ID
+// (associated with the request) lazily only when required.
```
Suggestion:

lazyAuditID implements Stringer interface to lazily retrieve the audit ID associated with the request

tkashem commented:
/retest

tkashem commented:
/retest

neolit123 commented:
/remove-area kubeadm
/remove-sig cluster-lifecycle

p0lyn0mial commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:
```diff
@@ -201,6 +202,12 @@ func newRequestForProxy(location *url.URL, req *http.Request) (*http.Request, co
 
+	// If the original request has an audit ID, let's make sure we propagate this
+	// to the aggregated server.
+	if auditID, found := genericapirequest.AuditIDFrom(req.Context()); found {
```
have you checked `newReq.Header = utilnet.CloneHeader(req.Header)` ?

`req` should already have the `AuditID` header set by the filter

tkashem replied:
If the user specify an audit ID only then `req` will already have the `Audit-ID` header. If we generate an audit ID then we save it in the context. 
if the context already has an audit ID then it has higher precedence, and we want to set it to the new `req`.

p0lyn0mial replied:
okay, makes sense, thx.

tkashem commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -52,14 +53,11 @@ func NewEventFromRequest(req *http.Request, requestReceivedTimestamp time.Time,
-		ev.AuditID = types.UID(ids)
-	} else {
-		ev.AuditID = types.UID(uuid.New().String())
+	auditID, found := request.AuditIDFrom(req.Context())
```
yes, it is set when the builder chain has `WithAuditID` filter. Here, it is just handling an error case.

p0lyn0mial replied:
could you point me to the error case?

tkashem replied:
by `error case` i meant handling the case where `found` is false, instead of doing `auditID, _ := request.AuditIDFrom(req.Context())`

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+func withAuditID(handler http.Handler, newAuditIDFunc func() string) http.Handler {
+	if newAuditIDFunc == nil {
```
it should not be `nil`, here it is just handling an error path. Maybe I can print an error log message here.

sttts commented:
/lgtm
/approve
